### PR TITLE
Fix dotnet path for RoslynCodeTaskFactory

### LINKED ISSUE #9052: API-using application is spawned instead of `dotnet` when using CodeTaskFactory
I'm trying to open csproj files via `Microsoft.CodeAnalysis.MSBuild.MSBuildWorkspace.OpenProjectAsync` from my console app. One of the projects, a rather big one, hangs the execution and spawns another instance of the app.

Tried both v4.6.0 from nuget.org and built myself from https://github.com/dotnet/roslyn/releases/tag/Visual-Studio-2022-Version-17.6.4, same behavior.

Tried to debug but quickly got lost... My best guess is that MSBuild is trying to create additional processes (for compilation?) to parallelize since it's a large project, but instead instantiates my app.
Any suggestions much appreciated.

### PR DESCRIPTION
**Pull request by GangWang01**
Fixes #9052

### Context
For RoslynCodeTaskFactory, dotnet path is wrongly set using the path of the application when running the application via application.exe.

### Changes Made
Get dotnet path through the following in order.

1. Honor the environment variable DOTNET_HOST_PATH for tools and MSBuild tasks within the SDK to ensure a consistent experience.
2. Fallback to get dotnet path from current process which might be dotnet executable.
3. If the path resolved above is not dotnet executable, rely on dotnet via the system's PATH.

### Testing
Add a test that verifies build the task of RoslynCodeTaskFactory using API.

### CODE REVIEW

File: src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs [hunk 1]
```diff
@@ -1068,6 +1068,61 @@ public override bool Execute()
...
+
+  <UsingTask
+    TaskName=""Custom1""
+    TaskFactory=""RoslynCodeTaskFactory""
+    AssemblyFile=""$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll"" >
+    <ParameterGroup>
+      <SayHi ParameterType=""System.String"" Required=""true"" />
+    </ParameterGroup>
+    <Task>
+      <Reference Include=""{typeof(Enumerable).Assembly.Location}"" />
```
Comment (rainersigwald) [inline src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs:1085]: You shouldn't need this line--do you?

File: src/Tasks/RoslynCodeTaskFactory/RoslynCodeTaskFactoryCompilers.cs [hunk 2]
```diff
@@ -99,7 +120,8 @@ protected override string GenerateFullPathToTool()
             }
 
 #if RUNTIME_TYPE_NETCORE
-            return DotnetCliPath;
+            Log.LogMessageFromText($"dotnet path is {dotnetCliPath}", StandardOutputImportanceToUse);
```
Comment (rainersigwald) [inline src/Tasks/RoslynCodeTaskFactory/RoslynCodeTaskFactoryCompilers.cs:123]: Feels like this should be at most low-pri, and I question having it at all. Do we not log the command line we're going to invoke before we invoke it? If we do is that clear enough?

File: src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs [hunk 3]
```diff
@@ -1068,6 +1068,61 @@ public override bool Execute()
...
+
+            var project = env.CreateTestProjectWithFiles("p1.proj", text);
+
+            var logger = project.BuildProjectExpectSuccess();
+#if !FEATURE_RUN_EXE_IN_TESTS
+            var filter = "dotnet path is ";
+#else
+            var filter = "Compiling task source code";
+
+#endif
```
Comment (rainersigwald) [inline src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs:1114]: I don't think I understand the `#if` here. Can you explain why the behavior is conditional? Should we make the test netcore-only?

File: src/Tasks/RoslynCodeTaskFactory/RoslynCodeTaskFactoryCompilers.cs [hunk 4]
```diff
@@ -14,7 +15,7 @@ namespace Microsoft.Build.Tasks
     internal abstract class RoslynCodeTaskFactoryCompilerBase : ToolTaskExtension
     {
 #if RUNTIME_TYPE_NETCORE
-        private static readonly string DotnetCliPath = System.Diagnostics.Process.GetCurrentProcess().MainModule.FileName;
+        private readonly string dotnetCliPath;
```
Comment (rainersigwald) [inline src/Tasks/RoslynCodeTaskFactory/RoslynCodeTaskFactoryCompilers.cs:18]: our convention is that `private` fields should start with an `_`

```suggestion
        private readonly string _dotnetCliPath;
```

(we don't do this universally everywhere but it's used in this file, for example `_executablePath`, so let's stick with it here)

File: src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs [hunk 5]
```diff
@@ -1068,6 +1068,49 @@ public override bool Execute()
...
+            using var env = TestEnvironment.Create();
+            RunnerUtilities.ApplyDotnetHostPathEnvironmentVariable(env);
+            var dotnetPath = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH");
+
+            var project = env.CreateTestProjectWithFiles("p1.proj", text);
+            var logger = project.BuildProjectExpectSuccess();
+            var logLines = logger.AllBuildEvents.Select(a => a.Message);
+            var log = string.Join("\n", logLines);
+            var messages = logLines.Where(l => l.Contains(dotnetPath)).ToList();
+            messages.Count.ShouldBe(1, log);
```
Comment (rainersigwald) [inline src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs:1110]: Nit: with Shouldly a failure here would be easier to understand as

```suggestion
            logLines.Where(l => l.Contains(dotnetPath)).Count.ShouldBe(1, log);
```

This also avoids creating a concrete `List<T>`, which isn't generally required unless we're going to enumerate it more than once.

**Review by rainersigwald — COMMENTED** [review]
Just a couple of nits, but looks good!

File: src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs (hunk 1 above)
Comment (GangWang01) [reply src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs]: It's not needed. Removed.

File: src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs (hunk 3 above)
Comment (GangWang01) [reply src/Tasks.UnitTests/RoslynCodeTaskFactory_Tests.cs]: Agree. The fix is for .NET version of msbuild. It's not necessary to test .net framework.

**Review by YuliiaKovalova — APPROVED** [review]

**Review by rainersigwald — APPROVED** [review]

**Pull request merged**